File: numass-core/src/main/java/inr/numass/data/storage/NumassStorage.java
/* 
 * Copyright 2015 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package inr.numass.data.storage;

import hep.dataforge.context.Context;
import hep.dataforge.events.Event;
import hep.dataforge.events.EventBuilder;
import hep.dataforge.exceptions.StorageException;
import hep.dataforge.meta.Meta;
import hep.dataforge.storage.filestorage.FileStorage;
import inr.numass.data.api.NumassSet;
import inr.numass.data.legacy.NumassDatFile;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ByteChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import static java.nio.file.StandardOpenOption.CREATE;
import static java.nio.file.StandardOpenOption.WRITE;


/**
 * The file storage containing numass data directories or zips.
 * <p>
 * Any subdirectory is treated as numass data directory. Any zip must have
 * {@code NUMASS_ZIP_EXTENSION} extension to be recognized. Any other files are
 * ignored.
 * </p>
 *
 * @author Alexander Nozik
 */
public class NumassStorage extends FileStorage {

    public static final String NUMASS_ZIP_EXTENSION = ".nm.zip";
    public static final String NUMASS_DATA_LOADER_TYPE = "numassData";

    protected NumassStorage(FileStorage parent, String path, Meta config) throws StorageException {
        super(parent, path, config);
        super.refresh();
    }

    public NumassStorage(Context context, Meta config) throws StorageException {
        super(context, config);
        super.refresh();
    }

    @Override
    protected void updateDirectoryLoaders() {
        try {
            this.loaders.clear();
            Files.list(getDataDir()).forEach(file -> {
                try {
                    if (Files.isDirectory(file)) {
                        Path metaFile = file.resolve(NumassDataLoader.META_FRAGMENT_NAME);
                        if (Files.exists(metaFile)) {
                            this.loaders.put(entryName(file),
                                    NumassDataLoader.fromDir(this, file, null));
                        } else {
                            this.shelves.put(entryName(file),
                                    new NumassStorage(this, entryName(file), meta()));
                        }
                    } else if (file.getFileName().endsWith(NUMASS_ZIP_EXTENSION)) {
                        this.loaders.put(entryName(file), NumassDataLoader.fromFile(this, file));
//                    } else if (file.getFileName().endsWith(".points")) {
//                        try {
//                            loaders.put(getFileName(file),
//                                    FilePointLoader.fromFile(this, file, true));
//                        } catch (Exception ex) {
//                            getLogger().error("Failed to build numass point loader from file {}", file.getName());
//                        }
                    } else {
                        //updating non-numass loader files
                        updateFile(file);
                    }
                } catch (IOException ex) {
                    LoggerFactory.getLogger(getClass()).error("Error while creating numass loader", ex);
                } catch (StorageException ex) {
                    LoggerFactory.getLogger(getClass()).error("Error while creating numass group", ex);
                }
            });
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
    }

    public void pushNumassData(String path, String fileName, ByteBuffer data) throws StorageException {
        if (path == null || path.isEmpty()) {
            pushNumassData(fileName, data);
        } else {
            NumassStorage st = (NumassStorage) buildShelf(path);
            st.pushNumassData(fileName, data);
        }
    }

    /**
     * Read nm.zip content and write it as a new nm.zip file
     *
     * @param fileName
     */
    @SuppressWarnings("unchecked")
    public void pushNumassData(String fileName, ByteBuffer data) throws StorageException {
        //FIXME move zip to internal
        try {
            Path nmFile = getDataDir().resolve(fileName + NUMASS_ZIP_EXTENSION);
            if (Files.exists(nmFile)) {
                LoggerFactory.getLogger(getClass()).warn("Trying to rewrite existing numass data file {}", nmFile.toString());
            }
            try (ByteChannel channel = Files.newByteChannel(nmFile, CREATE, WRITE)) {
                channel.write(data);
            }

            dispatchEvent(NumassDataPointEvent.build(getName(), fileName, (int) Files.size(nmFile)));
        } catch (IOException ex) {
            throw new StorageException(ex);
        }
    }

    @Override
    public NumassStorage createShelf(String path, Meta meta) throws StorageException {
        return new NumassStorage(this, path, meta);
    }

    /**
     * A list of legacy DAT files in the directory
     *
     * @return
     */
    public List<NumassSet> legacyFiles() {
        try {
            List<NumassSet> files = new ArrayList<>();
            Files.list(getDataDir()).forEach(file -> {
                if (Files.isRegularFile(file) && file.getFileName().toString().toLowerCase().endsWith(".dat")) {
                    String name = file.getFileName().toString();
                    try {
                        files.add(new NumassDatFile(file, Meta.empty()));
                    } catch (Exception ex) {
                        LoggerFactory.getLogger(getClass()).error("Error while reading legacy numass file " + file.getFileName(), ex);
                    }
                }
            });
            return files;
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
    }

    public String getDescription() {
        return meta().getString("description", "");
    }

    public static class NumassDataPointEvent extends Event {

        public static final String FILE_NAME_KEY = "fileName";
        public static final String FILE_SIZE_KEY = "fileSize";

        public NumassDataPointEvent(Meta meta) {
            super(meta);
        }

        public static NumassDataPointEvent build(String source, String fileName, int fileSize) {
            return new NumassDataPointEvent(builder(source, fileName, fileSize).buildEventMeta());
        }

        public static EventBuilder builder(String source, String fileName, int fileSize) {
            return EventBuilder.make("numass.storage.pushData")
                    .setSource(source)
                    .setMetaValue(FILE_NAME_KEY, fileName)
                    .setMetaValue(FILE_SIZE_KEY, fileSize);
        }

        public int getFileSize() {
            return meta().getInt(FILE_SIZE_KEY, 0);
        }

        public String getFileName() {
            return meta().getString(FILE_NAME_KEY);
        }

        @Override
        public String toString() {
            return String.format("(%s) [%s] : pushed numass data file with name '%s' and size '%d'",
                    time().toString(), sourceTag(), getFileName(), getFileSize());
        }

    }

}

File: numass-core/src/main/kotlin/inr/numass/data/storage/NumassStorage.kt
/*
 * Copyright 2015 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package inr.numass.data.storage

import hep.dataforge.context.Context
import hep.dataforge.context.Global
import hep.dataforge.events.Event
import hep.dataforge.events.EventBuilder
import hep.dataforge.meta.Meta
import hep.dataforge.storage.StorageElement
import hep.dataforge.storage.files.FileStorage
import hep.dataforge.storage.files.FileStorageElement
import inr.numass.NumassEnvelopeType
import java.nio.file.Files
import java.nio.file.Path

/**
 * Numass storage directory. Works as a normal directory, but creates a numass loader from each directory with meta
 */
class NumassDirectory : FileStorage.Directory() {
    override val name: String = NUMASS_DIRECTORY_TYPE

    override suspend fun read(context: Context, path: Path, parent: StorageElement?): FileStorageElement? {
        val meta = FileStorage.resolveMeta(path){ NumassEnvelopeType.infer(it)?.reader?.read(it)?.meta }
        return if (Files.isDirectory(path) && meta != null) {
            NumassDataLoader(context, parent, path.fileName.toString(), path)
        } else {
            super.read(context, path, parent)
        }
    }

    companion object {
        val INSTANCE = NumassDirectory()
        const val NUMASS_DIRECTORY_TYPE = "inr.numass.storage.directory"

        /**
         * Simple read for scripting and debug
         */
        fun read(context: Context = Global, path: String): FileStorageElement?{
            return runBlocking { INSTANCE.read(context, context.getDataFile(path).absolutePath)}
        }
    }
}

class NumassDataPointEvent(meta: Meta) : Event(meta) {

    val fileSize: Int = meta.getInt(FILE_SIZE_KEY, 0)

    val fileName: String = meta.getString(FILE_NAME_KEY)

    override fun toString(): String {
        return String.format("(%s) [%s] : pushed numass data file with name '%s' and size '%d'",
                time().toString(), sourceTag(), fileName, fileSize)
    }

    companion object {


        const val FILE_NAME_KEY = "fileName"
        const val FILE_SIZE_KEY = "fileSize"

        fun build(source: String, fileName: String, fileSize: Int): NumassDataPointEvent {
            return NumassDataPointEvent(builder(source, fileName, fileSize).buildEventMeta())
        }

        fun builder(source: String, fileName: String, fileSize: Int): EventBuilder<*> {
            return EventBuilder.make("numass.storage.pushData")
                    .setSource(source)
                    .setMetaValue(FILE_NAME_KEY, fileName)
                    .setMetaValue(FILE_SIZE_KEY, fileSize)
        }
    }

}

//
///**
// * The file storage containing numass data directories or zips.
// *
// *
// * Any subdirectory is treated as numass data directory. Any zip must have
// * `NUMASS_ZIP_EXTENSION` extension to be recognized. Any other files are
// * ignored.
// *
// *
// * @author Alexander Nozik
// */
//class NumassStorage : FileStorage {
//
//    val description: String
//        get() = meta.getString("description", "")
//
//    private constructor(parent: FileStorage, config: Meta, shelf: String) : super(parent, config, shelf)
//
//    constructor(context: Context, config: Meta, path: Path) : super(context, config, path)
//
//    init {
//        refresh()
//    }
//
//    override fun refresh() {
//        try {
//            this.shelves.clear()
//            this.loaders.clear()
//            Files.list(dataDir).forEach { file ->
//                try {
//                    if (Files.isDirectory(file)) {
//                        val metaFile = file.resolve(NumassDataLoader.META_FRAGMENT_NAME)
//                        if (Files.exists(metaFile)) {
//                            this.loaders[entryName(file)] = NumassDataLoader.fromDir(this, file)
//                        } else {
//                            this.shelves[entryName(file)] = NumassStorage(this, meta, entryName(file))
//                        }
//                    } else if (file.fileName.endsWith(NUMASS_ZIP_EXTENSION)) {
//                        this.loaders[entryName(file)] = NumassDataLoader.fromFile(this, file)
//                    } else {
//                        //updating non-numass loader files
//                        updateFile(file)
//                    }
//                } catch (ex: IOException) {
//                    LoggerFactory.getLogger(javaClass).error("Error while creating numass loader", ex)
//                } catch (ex: StorageException) {
//                    LoggerFactory.getLogger(javaClass).error("Error while creating numass group", ex)
//                }
//            }
//        } catch (ex: IOException) {
//            throw RuntimeException(ex)
//        }
//
//    }
//
//    @Throws(StorageException::class)
//    fun pushNumassData(path: String?, fileName: String, data: ByteBuffer) {
//        if (path == null || path.isEmpty()) {
//            pushNumassData(fileName, data)
//        } else {
//            val st = buildShelf(path) as NumassStorage
//            st.pushNumassData(fileName, data)
//        }
//    }
//
//    /**
//     * Read nm.zip content and write it as a new nm.zip file
//     *
//     * @param fileName
//     */
//    @Throws(StorageException::class)
//    fun pushNumassData(fileName: String, data: ByteBuffer) {
//        //FIXME move zip to internal
//        try {
//            val nmFile = dataDir.resolve(fileName + NUMASS_ZIP_EXTENSION)
//            if (Files.exists(nmFile)) {
//                LoggerFactory.getLogger(javaClass).warn("Trying to rewrite existing numass data file {}", nmFile.toString())
//            }
//            Files.newByteChannel(nmFile, CREATE, WRITE).use { channel -> channel.write(data) }
//
//            dispatchEvent(NumassDataPointEvent.build(name, fileName, Files.size(nmFile).toInt()))
//        } catch (ex: IOException) {
//            throw StorageException(ex)
//        }
//
//    }
//
//    @Throws(StorageException::class)
//    override fun createShelf(shelfConfiguration: Meta, shelfName: String): NumassStorage {
//        return NumassStorage(this, shelfConfiguration, shelfName)
//    }
//
//    /**
//     * A list of legacy DAT files in the directory
//     *
//     * @return
//     */
//    fun legacyFiles(): List<NumassSet> {
//        try {
//            val files = ArrayList<NumassSet>()
//            Files.list(dataDir).forEach { file ->
//                if (Files.isRegularFile(file) && file.fileName.toString().toLowerCase().endsWith(".dat")) {
//                    //val name = file.fileName.toString()
//                    try {
//                        files.add(NumassDatFile(file, Meta.empty()))
//                    } catch (ex: Exception) {
//                        LoggerFactory.getLogger(javaClass).error("Error while reading legacy numass file " + file.fileName, ex)
//                    }
//
//                }
//            }
//            return files
//        } catch (ex: IOException) {
//            throw RuntimeException(ex)
//        }
//
//    }
//
//    @Throws(Exception::class)
//    override fun close() {
//        super.close()
//        //close remote file system after use
//        try {
//            dataDir.fileSystem.close()
//        } catch (ex: UnsupportedOperationException) {
//
//        }
//
//    }
//

//
//    companion object {
//
//        const val NUMASS_ZIP_EXTENSION = ".nm.zip"
//        const val NUMASS_DATA_LOADER_TYPE = "numassData"
//    }
//
//}

//class NumassStorageFactory : StorageType {
//
//    override fun type(): String {
//        return "numass"
//    }
//
//    override fun build(context: Context, meta: Meta): Storage {
//        if (meta.hasValue("path")) {
//            val uri = URI.create(meta.getString("path"))
//            val path: Path
//            if (uri.scheme.startsWith("ssh")) {
//                try {
//                    val username = meta.getString("userName", uri.userInfo)
//                    //String host = meta.getString("host", uri.getHost());
//                    val port = meta.getInt("port", 22)
//                    val env = SFTPEnvironment()
//                            .withUsername(username)
//                            .withPassword(meta.getString("password", "").toCharArray())
//                    val fs = FileSystems.newFileSystem(uri, env, context.classLoader)
//                    path = fs.getPath(uri.path)
//                } catch (e: Exception) {
//                    throw RuntimeException(e)
//                }
//
//            } else {
//                path = Paths.get(uri)
//            }
//            if(!Files.exists(path)){
//                context.logger.info("File $path does not exist. Creating a new storage directory.")
//                Files.createDirectories(path)
//            }
//            return NumassStorage(context, meta, path)
//        } else {
//            context.logger.warn("A storage path not provided. Creating default root storage in the working directory")
//            return NumassStorage(context, meta, context.workDir)
//        }
//    }
//
//    companion object {
//
//        /**
//         * Build local storage with Global context. Used for tests.
//         *
//         * @param file
//         * @return
//         */
//        fun buildLocal(context: Context, file: Path, readOnly: Boolean, monitor: Boolean): FileStorage {
//            val manager = context.load(StorageManager::class.java, Meta.empty())
//            return manager.buildStorage(buildStorageMeta(file.toUri(), readOnly, monitor)) as FileStorage
//        }
//
//        fun buildLocal(context: Context, path: String, readOnly: Boolean, monitor: Boolean): FileStorage {
//            val file = context.dataDir.resolve(path)
//            return buildLocal(context, file, readOnly, monitor)
//        }
//
//        fun buildStorageMeta(path: URI, readOnly: Boolean, monitor: Boolean): MetaBuilder {
//            return MetaBuilder("storage")
//                    .setValue("path", path.toString())
//                    .setValue("type", "numass")
//                    .setValue("readOnly", readOnly)
//                    .setValue("monitor", monitor)
//        }
//    }
//}
